relax strict dtype checking in ecsv to a warning

emirkmo opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

**Edit: Needs to wait for #12880 due to likely Merge conflicts.**

This pull request is to address strict type checking in ecsv.py which prevents `BinnedTimeSeries` (and likely other) classes from reading and writing non-standard dtypes to and from an ecsv file. Instead of raising an exception, we issue a `UserWarning` instead, and attempt to go through the file reading. If this fails, the supported dtypes are reported in the warning so the user may switch to them instead.

Also updated tests to expect a warning but I am not sure if this was the right way to approach it. (`test_read_bad_datatype`, `test_read_complex`). 

`UserWarning` text is now:   
`(f'unexpected datatype {col.dtype!r} of column {col.name!r} '
                       f'is not in default datatypes {ECSV_DATATYPES}, using anyway.')`

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Edit: Fixes #12885 
~~Fixes #12840~~

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

emirkmo commented:
@taldcroft Tagging you here to show my proposal. Let me know if this is not what you had in mind or I am missing something. Don't want to break the much more widely used Table.read functionality on accident.

taldcroft commented:
Excellent this is on the right track. I'll try to get back later for a review.

emirkmo commented:
Testing failures seem unrelated, might be due to #12823 . At least I hope.

taldcroft commented:
The only quick thought I had was whether it would be worth making a subclass like `InvalidEcsvDatatype(AstropyUserWarning)` that would make it easier to squelch this warning globally.

emirkmo commented:
I thought to make a custom warning but didn’t see many of them in the warnings module, so figured it wasn’t widely used in astropy. Can we make the warning inside ecsv.py by inheriting from UserWarning or should it go into to warnings module?

pep8speaks commented:
Hello @emirkmo :wave:! It looks like you've made some changes in your pull request, so I've checked the code again for style.


* In the file [`astropy/io/ascii/ecsv.py`](https://github.com/astropy/astropy/blob/50c049543de77cac169a89343c0214455a349487/astropy/io/ascii/ecsv.py):

> [Line 27:1](https://github.com/astropy/astropy/blob/50c049543de77cac169a89343c0214455a349487/astropy/io/ascii/ecsv.py#L27): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

emirkmo commented:
@taldcroft Okay, I added InvalidEcsvDatatypeWarning as and changed the tests. Let me know if this is looking alright to you.

taldcroft commented:
Also I'd request squashing down to one commit since this is a pretty small and contained change.

emirkmo commented:
Ok, I think I squashed but I wasn't really sure how to do it with your commit having the credentials. please let me know if it looks ok. There are testing failures in py3.9 related to warnings and exceptions having None I think. Thanks for the assistance!

taldcroft commented:
@emirkmo - something we wrong with CI. This may be fixed by rebasing on main and force pushing. Thanks for your patience...

emirkmo commented:
@taldcroft Ok, I rebased. I also added a test for str instead of str raising a warning, but feel free to reject that commit or let me know if I should drop it (or rebase it). Like I said, tests are passing locally for me, but I only re-tested the io.ascii package

emirkmo commented:
Alright, I found the error. Older versions of numpy (`1.18.*`) do not support giving `'complex'` as a datatype string. Needs to be `1.19.*`

```
vals = ['1+1j', '2+2j']
numpy.array(vals, 'complex')
```
fails on numpy=1.18.1 and 1.18.5, but passes on >=1.19.* which is in the tests. And this comes from `test_read_complex()` because we now give a warning and proceed instead of raising an error. Should the tests be modified to expect a warning AND a (possible) error?

emirkmo commented:
@taldcroft All tests passing now, so we can merge, I hope! 

See my comment above re: tests. I'm also squashing to 1 commit as you asked and changing changelog from bugfix to api change. (will re-run CI, but it already passed before the squash).

While I wish this would go in as a bugfix I see why it should wait for next minor version. Thanks for the help.

Finally, possibly, this behavior should be documented in the [docs](https://docs.astropy.org/en/stable/io/ascii/ecsv.html) somewhere.

pllim commented:
#12880 is merged. Please rebase. Thanks!

emirkmo commented:
I put my proposal under a new commit because I think the warning should be given even for V0.9 files.

I also simplified the tests as a result, removing V0.9 specific versions, and adding some asserts to check that the output after proceeding with warning is consistent with what happened with V0.9 (copying from the V0.9 tests)

taldcroft commented on astropy/io/ascii/tests/test_ecsv.py:
```diff
@@ -631,94 +629,80 @@ def test_read_not_json_serializable():
+                      match="unexpected datatype 'object' of column 'a' is not in default datatypes"):
+        t = Table.read(txt, format='ascii.ecsv')
+    assert t['a'][0] == "fail"
+    assert type(t['a'][1]) is str
```
Yeah, so this is part of the original rationale for being strict, namely that this "object" type doesn't really round trip in the expected way. Anyway, the warning should do the trick.

taldcroft commented on astropy/io/ascii/ecsv.py:
```diff
@@ -175,13 +180,13 @@ def get_cols(self, lines):
+            # like datetime64 or object or python str, which are not in the ECSV standard.
+            if col.dtype not in ECSV_DATATYPES:
+                msg = (f'unexpected datatype {col.dtype!r} of column {col.name!r} '
+                       f'is not in default datatypes {ECSV_DATATYPES}, using anyway.')
```
The "unexpected results" mostly applies to the `object` type but I think it's a good idea to call out that they need to be careful.
```suggestion
                       f'is not in allowed ECSV datatypes {ECSV_DATATYPES}. '
                       'Using anyway as a numpy dtype but beware since unexpected results '
                       'are possible.)
```

taldcroft reviewed (commented):
@emirkmo - thanks, looks good with just minor comments.

emirkmo commented:
I implemented your minor comments and now all tests are passing @taldcroft. I'm going to squash as well, then I guess it's ready to merge.

emirkmo commented:
Is there anything new that needs to be done in terms of documentation? Also, I will fix the merge conflict.

taldcroft reviewed (approved):
Looks good @emirkmo , thanks!!

taldcroft commented:
@emirkmo - ping me after you fix the merge conflict.

emirkmo commented:
@taldcroft Merge conflicts resolved via rebase.

taldcroft reviewed (approved):
Thanks @emirkmo !
